core/ir/decode.h

```diff
@@ -41,6 +41,7 @@
 #define DECODE_H
```

Comment:
@abhinav92003 and I discussed this one.

+ We assume PIC/PIE is a requirement for your use case and building a version of the library w/o relocations is not a possible solution?

+ Are there possible compiler optimizations to help here: I guess duplicate string literal and string shared suffix optimizations don't help much and we'd need the compiler to convert our pointer uses into offsets.

+ The code readability impact here isn't too bad; mostly it's a maintenance burden as now a script has to be run.  The script would be run as part of the build process, or a dev would run it manually whenever adding new entries?

+ > but affects some fake internal names (e.g. "(bad)" is converted to "BAD").

  Could we have some indication that `"BAD"` is not really an opcode name?  `"_BAD_"`?

Reply:
Worth getting feedback from @khuey as well

Reply:
> We assume PIC/PIE is a requirement for your use case and building a version of the library w/o relocations is not a possible solution?

Yes, we need to support all build modes.

> Are there possible compiler optimizations to help here

I can't think of any. The main problem is with relocations rather than deduping/etc.

> The script would be run as part of the build process, or a dev would run it manually whenever adding new entries?

The intention is that it will run automatically, but I din't look into how easy it is to integrate into cmake, I assume should have required support.

> Could we have some indication that "BAD" is not really an opcode name? "_BAD_"?

We sure can adjust them any way you want (and these will also be 100% chanable in future patches).

Reply:
One other option is to make name `char[16]` (embed right in the struct).
Longest real x86 instruction I see are vpbroadcastmw2d/vbroadcasti64x4/vscatterpf1dpd. So 16 should be enough.
The compiler won't ensure that it's 0-terminated, but it can be reliably verified in tests.

Reply:
> One other option is to make name char[16] (embed right in the struct).

I like that: but it doesn't save as much space?

If we have 6778 table entries: the original proposal removed the 24 bytes of relocs plus 6 bytes for pointer-to-2-byte-offset which is 6778*30 = 203340 == your 200K savings.

Here we'd go from the 8-byte pointer to a 16-byte string, but removing the 24 reloc bytes, for a net reduction of 6778*16 = 108448 == just 100K.  But the strings themselves disappear: so if the strings averaged 6 chars with newline we're saving 6778*22 = 149116.  Is 150K savings enough?

Reply:
Right. I did a quick prototype and this version saves 76K on top of char[16] version:

```
    FILE SIZE        VM SIZE    
 --------------  -------------- 
   +43% +18.9Ki   +44% +18.9Ki    .rodata
 -27.3% -96.1Ki -27.3% -96.1Ki    .data.rel.ro
  -7.8% -76.1Ki -14.1% -77.1Ki    TOTAL
```

Reply:
@khuey do you have any feedback/objections to this and #6211?